tools/emscripten.py

```diff
@@ -989,29 +950,58 @@ def create_receiving(function_exports, tag_exports):
   if not settings.DECLARE_ASM_MODULE_EXPORTS:
     return ''
 
-  if settings.MINIMAL_RUNTIME:
-    # Exports are assigned inside a function to variables
-    # existing in top level JS scope, i.e.
-    # var _main;
-    # function assignWasmExports(wasmExport) {
-    #   _main = wasmExports["_main"];
-    generate_dyncall_assignment = settings.DYNCALLS and '$dynCall' in settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE
-    exports = [x for x in function_exports if x != building.WASM_CALL_CTORS]
-    receiving.append('function assignWasmExports(wasmExports) {')
-    for s in exports:
-      mangled = asmjs_mangle(s)
-      dynCallAssignment = ('dynCalls["' + s.replace('dynCall_', '') + '"] = ') if generate_dyncall_assignment and mangled.startswith('dynCall_') else ''
-      should_export = settings.EXPORT_ALL or (settings.EXPORT_KEEPALIVE and mangled in settings.EXPORTED_FUNCTIONS)
-      export_assignment = ''
-      if settings.MODULARIZE and should_export:
-        export_assignment = f"Module['{mangled}'] = "
-      receiving.append(f"  {export_assignment}{dynCallAssignment}{mangled} = wasmExports['{s}'];")
-    receiving.append('}')
+  # Exports are assigned inside a function to variables
+  # existing in top level JS scope, i.e.
+  # var _main;
+  # function assignWasmExports(wasmExport) {
+  #   _main = wasmExports["_main"];
+  generate_dyncall_assignment = settings.MINIMAL_RUNTIME and settings.DYNCALLS and '$dynCall' in settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE
+  exports = {name: sig for name, sig in function_exports.items() if name != building.WASM_CALL_CTORS}
+
+  if settings.ASSERTIONS:
+    # In debug builds we generate trapping function in case
+    # folks try to call use a reference that was taked before the
+    # wasm module is available.
+    for sym in exports:
+      mangled = asmjs_mangle(sym)
+      if settings.MODULARIZE == 'instance' and should_export(mangled):
+        maybe_export = 'export '
+      else:
+        maybe_export = ''
+      receiving.append(f"{maybe_export}var {mangled} = makeInvalidEarlyAccess('{mangled}');")
+  elif settings.MODULARIZE == 'instance':
+    # In MODULARIZE == 'instance' mode we declare each export individually
```

Comment:
Alternatively, you could use `var a, b, ...` unconditionally in both branches, and for ESM integration mode emit `export {a, b};` in a separate step.

It would be functionally equivalent, but a bit simpler implementation-wise.

Reply:
Great idea! That simplifies that code nicely.